### PR TITLE
Generate massaged method names for Python protos

### PR DESCRIPTION
**Pull request by soltanmm**
We need to do this before GA; the plan is to add a `snake_case` method option in upstream protobuf (cc @haberman, @nathanielmanistaatgoogle) to override this massaging in cases where it may fail miserably (like where it ends up generating the same name for two different method names).

### CODE REVIEW

File: src/python/grpcio/tests/protoc_plugin/protos/service/test_service.proto [hunk 1]
```diff
@@ -62,3 +62,34 @@ service TestService {
...
+  rpc NAMETEST(SimpleRequest) returns (SimpleResponse);
+}
+
+service MassagedNames3 {
+  // -> 'name_test1'
+  rpc NAMETest1(SimpleRequest) returns (SimpleResponse);
+}
+
+service MassagedNames4 {
+  // -> 'name_tes_t1' (we *only* use capital letters to distinguish word
```
Comment (kpayson64) [inline src/python/grpcio/tests/protoc_plugin/protos/service/test_service.proto]: Is there a spec your going off for these rules?  I think the intuitive result for this name would be ```name_test1```

Comment (soltanmm) [inline src/python/grpcio/tests/protoc_plugin/protos/service/test_service.proto]: That would be an intuitive result (alternatively I went for `name_test_1` in an earlier iteration), but we don't have a spec.

Comment (kpayson64) [inline src/python/grpcio/tests/protoc_plugin/protos/service/test_service.proto]: There are probably some more esoteric edge cases, but my 2 cents is that sequential non-uppercase letters should never be split.

Comment (soltanmm) [inline src/python/grpcio/tests/protoc_plugin/protos/service/test_service.proto]: I *believe* that's currently followed by virtue of non-uppercase letters never being used to create a word boundary.

File: src/compiler/python_generator.cc [hunk 2]
```diff
@@ -170,6 +170,45 @@ grpc::string ModuleAlias(const grpc::string& filename) {
...
+  grpc::string result;
+  size_t word_start = 0;
+  size_t word_end = 0;
+  while (true) {
+    word_start = word_end;
+    if (word_start + 1 < method_name.size() &&
+        std::isupper(method_name[word_start + 1])) {
+      // the word starting at `word_start` is an all-caps acronym
+      for (word_end = word_start + 1; word_end < method_name.size(); ++word_end) {
+        if (!std::isupper(method_name[word_end])) {
```
Comment (kpayson64) [inline src/compiler/python_generator.cc]: What if we change this to ```if (std::islower())```.  Its not perfect, but I think it covers mixed caps/numerics slightly better.

File: src/compiler/python_generator.cc [hunk 3]
```diff
@@ -303,12 +344,13 @@ bool PrintBetaServerFactory(const grpc::string& package_qualified_service_name,
                                    &output_message_module_and_class)) {
         return false;
       }
+      grpc::string method_name = ToSnakeCase(method->name());
       method_implementation_constructors.insert(
-          make_pair(method->name(), method_implementation_constructor));
+          make_pair(method_name, method_implementation_constructor));
       input_message_modules_and_classes.insert(
```
Comment (kpayson64) [inline src/compiler/python_generator.cc]: I don't think we want to mangle the serialize/deserialize names, on the wire those should be the raw proto names.

File: src/compiler/python_generator.cc [hunk 4]
```diff
@@ -397,12 +439,13 @@ bool PrintBetaStubFactory(const grpc::string& package_qualified_service_name,
                                    &output_message_module_and_class)) {
         return false;
       }
+      grpc::string method_name = ToSnakeCase(method->name());
       method_cardinalities.insert(
-          make_pair(method->name(), method_cardinality));
+          make_pair(method_name, method_cardinality));
       input_message_modules_and_classes.insert(
-          make_pair(method->name(), input_message_module_and_class));
+          make_pair(method_name, input_message_module_and_class));
```
Comment (kpayson64) [inline src/compiler/python_generator.cc]: Same as above.  I *think* mangling the names for cardinalities is correct though.

**Comment by kpayson64** [comment]
Make sure to update the servicers/stubs for the QPS/Interop/Stress tests as well.

**Comment by soltanmm** [comment]
@nathanielmanistaatgoogle It doesn't seem possible to specify that RPC method names are different from the Python attribute names on the client side. Is that so?

**Comment by kpayson64** [comment]
This was a top ask from the API review.

**Comment by nicolasnoble** [comment]
Stale ?

**Comment by nathanielmanistaatgoogle** [comment]
@nicolasnoble: _very slowly_ under debate.

**Comment by soltanmm-google** [comment]
Ahahahahahaha.... hahaha........ haaaaahhhhhhh... we're past 1.0. This is *dead*.

**Comment by thelinuxfoundation** [comment]
Thank you for your pull request. Before we can look at your contribution, we need to ensure all contributors are covered by a Contributor License Agreement.

After the following items are addressed, please respond with a new comment here, and the automated system will re-verify.

* User @soltanmm-google isn't covered by a CLA. They will need to complete the form at https://identity.linuxfoundation.org/projects/cncf

_Regards,
CLA GitHub bot_
